pkg/sql/opt/norm/rules/join.opt

```diff
@@ -532,6 +532,28 @@ $left
     $private
 )
 
+#ConvertProjectionsToFiltersLeft
+[ConvertProjectionsToFiltersLeft, Normalize]
+(InnerJoin | InnerJoinApply | LeftJoin | LeftJoinApply | SemiJoin
+    $left:(Project)
```

Comment:
We try to put as much logic as possible in the rule itself. For example, we usually do `(Project $input:* $projections:* $passthrough:*)` and then just pass the input/projections/passthrough directly rather than having the custom function cast to ProjectExpr

Reply:
Yes- this simplifies things a lot - I've added a new commit to make this change

Reply:
Although rebasing the upstream changes from cockroachdb appears to have broken this pull request, so I will close this and reopen it